// vott-react/dist/lib/common/mockFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MockFactory {
    /**
     * Create array of fake ITag
     * @param count Number of tags
     */
    static createTestTags(count = 5) {
        const tags = [];
        for (let i = 0; i < count; i++) {
            tags.push(MockFactory.createTestTag(i.toString()));
        }
        return tags;
    }
    /**
     * Create fake ITag with random color
     * @param name Name of tag
     */
    static createTestTag(name = "Test Tag") {
        return {
            name: `Tag ${name}`,
            color: MockFactory.randomColor(),
        };
    }
    /**
     * Runs function that updates the UI, and flushes call stack
     * @param func - The function that updates the UI
     */
    static flushUi(func = null) {
        return new Promise((resolve) => {
            if (func) {
                func();
            }
            setImmediate(resolve);
        });
    }
    /**
     * Generates a random color string
     */
    static randomColor() {
        return [
            "#",
            MockFactory.randomColorSegment(),
            MockFactory.randomColorSegment(),
            MockFactory.randomColorSegment(),
        ].join("");
    }
    /**
     * Generates random color segment
     */
    static randomColorSegment() {
        const num = Math.floor(Math.random() * 255);
        return num.toString(16);
    }
}
exports.default = MockFactory;
